src/clients/BundleDataClient/utils/DataworkerUtils.ts

```diff
@@ -297,9 +306,59 @@ export function _buildPoolRebalanceRoot(
     });
   });
 
-  // Add to the running balance value from the last valid root bundle proposal for {chainId, l1Token}
-  // combination if found.
-  addLastRunningBalance(latestMainnetBlock, runningBalances, clients.hubPoolClient);
+  // Add to the running balance value from the last valid root bundle proposal. A root bundle proposal which is still in liveness and has not yet been executed counts as a valid proposal.
+  // If the mainnetBundleEndBlock for this pool rebalance root corresponds to the pending root bundle or some root bundle before the pending root bundle, then we can fetch running balances directly from `ExecutedRootBundle` events.
+  if (
+    !clients.hubPoolClient.hasPendingProposal() ||
+    clients.hubPoolClient.getPendingRootBundle()!.bundleEvaluationBlockNumbers[1] <= mainnetBundleEndBlock
+  ) {
+    addLastRunningBalance(latestMainnetBlock, runningBalances, clients.hubPoolClient);
+  } else {
+    // Otherwise, the pool rebalance root for this root bundle must be for an optimistic proposal, so we need to reconstruct the pool rebalance root for the pending root bundle to obtain the running balances.
+    // @dev It is safe to index the hub pool client's proposed root bundles here since there is guaranteed to be a pending proposal in this code block.
+    const mostRecentProposedRootBundle = clients.hubPoolClient.getLatestProposedRootBundle();
+    const blockRangesForChains = getImpliedBundleBlockRanges(
+      clients.hubPoolClient,
+      clients.configStoreClient,
+      mostRecentProposedRootBundle
+    );
+    // We are loading data from a pending root bundle, so we want to use arweave if possible.
+    const prevRootBundleData = await clients.bundleDataClient.loadData(
+      blockRangesForChains,
+      clients.spokePoolClients,
+      true
+    );
+    const prevPoolRebalanceRoot = await _buildPoolRebalanceRoot(
+      latestMainnetBlock,
+      blockRangesForChains[0][1],
+      prevRootBundleData.bundleDepositsV3,
+      prevRootBundleData.bundleFillsV3,
+      prevRootBundleData.bundleSlowFillsV3,
+      prevRootBundleData.unexecutableSlowFills,
+      prevRootBundleData.expiredDepositsToRefundV3,
+      clients,
+      maxL1TokenCountOverride
+    );
+    // Update the running balances by adding the pending root bundle's running balance amounts to this bundle's running balances.
+    const { runningBalances: prevRunningBalances } = prevPoolRebalanceRoot;
+    Object.keys(runningBalances).forEach((repaymentChainId) => {
+      Object.keys(runningBalances[Number(repaymentChainId)]).forEach((l1TokenAddress) => {
+        const updateRunningBalanceAmount = prevRunningBalances[Number(repaymentChainId)]?.[l1TokenAddress];
+        // If the pending root bundle didn't have a running balance, fall back to the last running balance for that token and chain pair.
+        const { runningBalance: fallbackRunningBalance } = clients.hubPoolClient.getRunningBalanceBeforeBlockForChain(
```

Comment:
Instead of precomputing all running balances for `prevPoolRebalanceRoot` by calling `await _buildPoolRebalanceRoot`, could we instead remove that call and just use the result of `hubPoolClient.getRunningBalanceBeforeBlockForChain` in order to set the `runningBalances` value?

I guess i'm saying that I don't think we need the `prevRunningBalances` value as you have written it here, since this value will be equal to `hubPoolClient.getRunningBalanceBeforeBlockForChain`if you pass in `latestMainnetBlock`.

Maybe I'm wrong here, just a thought.

Reply:
yeah It hink its helpful to have a function that spits out running balances just from all the `bundle` events, and then you can choose how to add the `lastRunningBalance`

Reply:
Yeah this is much simpler. I did that here https://github.com/across-protocol/sdk/pull/1142/commits/f4d7d7284e6aff9ab83c61a2ff318ff025f02fd6